src/colors/conversions.jl

```diff
@@ -0,0 +1,105 @@
+"""
+    channelify(m::AbstractArray{CT,N}) where {CT <: Colorant, N}
+Channelify function.
+
+Input should be in ``WHN`` order ``\\implies (*, batch)``.
+Output will be in the order ``(*, channels, batch)``.
+
+# Examples
+```jldoctest; setup = :(using Images,DiffImages)
+julia> input_size = (16, 16, 2)
+(16, 16, 2)
+
+julia> size(channelify(HSV.(rand(BGR,input_size...))))
+(16, 16, 3, 2)
+```
+"""
+function channelify(m::AbstractArray{CT,N}) where {CT <: Colorant, N}
+    e = eltype(m)
+    m = channelview(m)
+    if e <: AbstractGray
+        m = reshape(m, 1, size(m)...)
+    end
+    t = ntuple(identity, ndims(m))
+    m = permutedims(m, (t[2:end-1]...,1,t[end]))
+    return m
+end
+
+@adjoint function channelview(x::AbstractArray{T,N}) where {T, N}
+    e = eltype(x)
+    y = channelview(x)
+    function pullback(Δ)
+        return (collect(colorview(e,Δ)),)
+    end
+    return (y, pullback)
+end
+
+@adjoint function colorview(T, x)
+    y = colorview(T,x)
+    function pullback(Δ)
+        return (nothing, channelview(Δ))
+    end
+    return (y, pullback)
+end
+
+# adjoint for (::Colorant{T,N})(x::Real)
+for f in (:HSV,:AHSV,:HSVA,
+          :Gray,:AGray,:GrayA,
+          :HSL,:AHSL,:HSLA,
+          :RGB,:ARGB,:RGBA,
+          :BGR,:ABGR,:BGRA,
+          :XYZ,:AXYZ,:XYZA,
+          :xyY,:AxyY,:xyYA,
+          :Lab,:ALab,:LabA,
+          :Luv,:ALuv,:LuvA,
+          :LCHab,:ALCHab,:LCHabA,
+          :LCHuv,:ALCHuv,:LCHuvA,
+          :DIN99,:ADIN99,:DIN99A,
+          :LMS,:ALMS,:LMSA,
+          :YIQ,:AYIQ,:YIQA)
+    @eval @adjoint function $f(x::Real...)
+        y = $f(x::Real...)
+        function pull(Δ...)
+            return (Δ...,)
+        end
+        return (y, pull)
+    end
+end
+
+# Constructor adjoint
+# function ChainRules.rrule(::Type{HSL{T}}, x, y, z) where T <: AbstractFloat
+#     β = HSL{T}(x, y, z)
+#     function Lab_pullback(Δ)
+#         @show Δ x y z β
+#         return (ChainRules.NoTangent(),Δ.h,Δ.s,Δ.l)
+#     end
+#     return (β, Lab_pullback)
+# end
+
+"""
+    colorify(color::Type{CT}, m::AbstractArray) where CT <: Colorant
+Colorify function.
+
+Expecting an input of the type ``(*, channels, batch)``.
+Converts the array to the `color` specified.
+
+
+# Examples
+```jldoctest; setup = :(using Images,DiffImages)
+julia> input_size = (25, 25, 3, 7)
+(25, 25, 3, 7)
+
+julia> size(colorify(HSV, rand(input_size...)))
+(25, 25, 7)
+```
+"""
+function colorify(color::Type{CT}, m::AbstractArray) where CT <: Colorant
```

Comment:
I feel like we may want to restrict this function to `AbstractArray{T,4}` since the case of `(28,28,3)` with Grayscale should be a simple reinterpret

Reply:
The current function gives the freedom of any number of dims >= 2. We could restrict it to 4 but it could be we may require something for higher dimensions in the future.

Reply:
My main concern is the case I highlighted earlier. If that is handled correctly, then I'm fine

Reply:
I understand that classic deep learning mainly deals with 4D tensor but I'm not very positive about this type constraint; especially when it's not due to implementation restriction.

Reply:
I am the first person to not restrict types for sure, just wanted to ensure we're doing the correct thing for different inputs we can expect to get.